File: capibara-website/app/models/Typedef.scala
package models
import play.api.libs.json._

object TypedefRef {
  implicit val typedefRefFormatter = Json.format[TypedefRef];
}

object Typedef {
  implicit val typedefFormatter = Json.format[Typedef];
}

case class Typedef(
    name: String,
    header: HeaderRef,
    summary: String,
    `type`: String,
    associated_ref: TypedefRef,
    description: String,
    os_affinity: Seq[String]
)

case class TypedefRef(
    `enum`: Option[Enum],
    struct: Option[Struct],
    none: Option[JsObject] // TypedefNone
) {
  require(
    `enum`.isDefined ^ struct.isDefined ^ none.isDefined,
    "one enum, struct, or none exclusively must be defined!"
  )
}

File: capibara-website/src/models/typedef.rs
use super::{_struct::Struct, enumeration::Enumeration, header::HeaderSummary};
use rocket::serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Clone)]
pub(crate) struct Typedef {
    pub name: Box<String>,
    pub header: HeaderSummary,
    pub summary: Box<String>,
    #[serde(rename = "type")]
    pub _type: Box<String>,
    pub associated_ref: TypedefRef,
    pub description: Box<String>,
    pub os_affinity: Vec<String>,
}

#[derive(Serialize, Deserialize, PartialEq, Clone)]
pub(crate) enum TypedefRef {
    #[serde(rename = "none")]
    None(None),
    #[serde(rename = "enum")]
    Enumeration(Enumeration),
    #[serde(rename = "struct")]
    Struct(Struct),
}

#[derive(Serialize, Deserialize, PartialEq, Clone)]
pub(crate) struct None {}
